lib/Dialect/AIEVec/Transforms/VectorToVectorConversions.cpp

```diff
@@ -326,6 +326,48 @@ struct HoistCastOpToDataSourcePattern : public RewritePattern {
   }
 };
 
+// This pattern swaps a UnaryOpA followed by UnaryOpB. This pattern can be used
+// to improve pattern matching for mixed-type arithmetic ops, by getting sign
+// extension ops closer to the single-type arithmetic operations.
+template <class UnaryOpA, class UnaryOpB>
+struct SwapUnaryOpsPattern : public RewritePattern {
+  // This function takes the chain of operations A->B, and returns the new type
+  // between B and A after the swap.
+  using InferTypeB2AFnTy = std::function<Type(UnaryOpA aOp, UnaryOpB bOp)>;
+  InferTypeB2AFnTy inferTypeB2A = nullptr;
+
+  SwapUnaryOpsPattern(MLIRContext *context, InferTypeB2AFnTy inferType)
+      : RewritePattern(UnaryOpB::getOperationName(), /*benefit=*/1, context),
+        inferTypeB2A(inferType) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    static_assert(
+        UnaryOpA::template hasTrait<OpTrait::OneOperand>(),
+        "SwapUnaryOps can only be instantiated for single-operand ops");
+    static_assert(
+        UnaryOpB::template hasTrait<OpTrait::OneOperand>(),
+        "SwapUnaryOps can only be instantiated for single-operand ops");
+    UnaryOpB bOp = cast<UnaryOpB>(op);
```

Comment:
Do we know op is UnaryOpB, should this be a dynamic_cast with a check?

Reply:
Can it be anything else? I believe that, as it is, this pattern should match only for UnaryOpB.

```c++
  SwapUnaryOpsPattern(MLIRContext *context, InferTypeB2AFnTy inferType)
      : RewritePattern(UnaryOpB::getOperationName(), /*benefit=*/1, context), inferTypeB2A(inferType) {}
```

But maybe I'm not understanding something?

Reply:
I guess `: public mlir::OpRewritePattern<UnaryOpB>` will work as well. So that you can avoid that cast?

Reply:
Yeah... I guess I was being lazy. Let me fix that 🙂

Reply:
Ok hadn't noticed the `RewritePattern(UnaryOpB::getOperationName()` but that looks like it would have worked too (I was looking for something like `: public mlir::OpRewritePattern<UnaryOpB>`) Thanks!